proc_macros/src/lib.rs

```diff
@@ -44,3 +45,108 @@ pub fn builder_args(input: TokenStream) -> TokenStream {
     }
     .into()
 }
+
+struct InvalidRuleTestSpec {
+    code: Expr,
+    errors: ExprArray,
+}
+
+impl Parse for InvalidRuleTestSpec {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let mut code: Option<Expr> = Default::default();
+        let mut errors: Option<ExprArray> = Default::default();
+        let content;
+        braced!(content in input);
+        while !content.is_empty() {
+            let key: Ident = content.parse()?;
+            content.parse::<Token![=>]>()?;
+            match &*key.to_string() {
+                "code" => {
+                    code = Some(content.parse()?);
+                }
+                "errors" => {
+                    errors = Some(content.parse()?);
+                }
+                _ => panic!("didn't expect key {}", key),
+            }
+            if !content.is_empty() {
+                content.parse::<Token![,]>()?;
+            }
+        }
+        Ok(Self {
+            code: code.expect("Expected 'code'"),
+            errors: errors.expect("Expected 'errors'"),
+        })
+    }
+}
+
+impl ToTokens for InvalidRuleTestSpec {
```

Comment:
I don't know if this is idiomatic but it felt like I should be able to "delegate" to something inside a `quote!()` for how to "print" this type

And it seemed like `quote::ToTokens` was that "something" (it's what `quote!()` calls on all of the interpolated things it looks like)